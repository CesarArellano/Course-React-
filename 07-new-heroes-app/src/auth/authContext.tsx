import { createContext } from "react";

export interface AuthState {
  user: {
    name: string;
    logged: boolean;
  }
  dispatch: React.Dispatch<any>
}

export const AuthContext = createContext({} as AuthState );