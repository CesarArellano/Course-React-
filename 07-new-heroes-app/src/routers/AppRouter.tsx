import { BrowserRouter, Route, Routes } from 'react-router-dom';

import { PrivateRoute } from './PrivateRoute';
import { LoginScreen } from '../components/login/LoginScreen';
import { PublicRoute } from './PublicRoute';
import { DashboardRoutes } from './DashboardRoutes';

export const AppRouter = () => {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/login" element={ 
          <PublicRoute>
            <LoginScreen />
          </PublicRoute>
          } 
        />
        <Route path="/*" element={ 
            <PrivateRoute>
              <DashboardRoutes />
            </PrivateRoute>
          }
        />
        {/* <Route path="/*" element={ <DashboardRoutes /> }/> */}
      </Routes>
    </BrowserRouter>
  );
}
