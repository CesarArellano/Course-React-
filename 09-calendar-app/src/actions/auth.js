import { fetchWithoutToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = ( email, password ) => {
  return async ( dispatch ) => {
    const resp = await fetchWithoutToken('auth', { email, password }, 'POST');
    const decodedData = await resp.json();
    
    if( decodedData.ok ) {
      localStorage.setItem( 'token', decodedData.token );
      localStorage.setItem( 'token-init-date', new Date().getTime() );
      dispatch( login({
        uid: decodedData.uid,
        name: decodedData.name,
      }));
    }
  }
};

const login = ( user ) => ({
  type: types.authLogin,
  payload: user,
});