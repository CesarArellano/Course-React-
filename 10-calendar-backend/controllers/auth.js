const { response } = require('express');
const User = require('../models/User');

const createUser = async ( req, res = response ) => {
  
  const { email, password } = req.body;
  try {
    let user = await User.findOne({ email })

    if( user ) {
      return res.status(400).json({
        ok: false,
        msg: 'El usuario ya existe con ese correo.'
      });
    }

    user = new User( req.body );
    user.save();
    
    res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name
    });

  } catch(error) {
    console.log(error);
    res.status(500).json({
      ok: false,
      msg: 'Please talk to the admin'
    });
  }
  
}

const userLogin = ( req, res = response ) => {
  const { email, password } = req.body;

  res.json({
    ok: true,
    msg: 'login',
    email,
    password
  });

};

const revalidateToken = ( req, res = response ) => {
  res.json({
    ok: true,
    msg: 'renew'
  });
};

module.exports = {
  createUser,
  userLogin,
  revalidateToken
};