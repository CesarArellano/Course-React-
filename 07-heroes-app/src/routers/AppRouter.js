import React, { useContext } from 'react'
import {
  BrowserRouter as Router,
  Switch
} from "react-router-dom";
import { AuthContext } from '../auth/AuthContext';

import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PublicRoute } from './PublicRoute';
import { PrivateRoute } from './PrivateRoute';

export const AppRouter = () => {
  const { user:{ logged } } = useContext(AuthContext);
  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute 
            exact 
            path="/login" 
            isAuthenticated={ logged }
            component={ LoginScreen } 
          />
          <PrivateRoute 
            path="/" 
            isAuthenticated={ logged }
            component={ DashboardRoutes } 
          />
        </Switch>
      </div>
    </Router>
  );
}
