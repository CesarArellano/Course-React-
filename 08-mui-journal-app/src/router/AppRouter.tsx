import { useSelector, useDispatch } from 'react-redux';
import { BrowserRouter, Navigate, Route, Routes } from 'react-router-dom'
import { AuthRoutes } from '../auth/routes/AuthRoutes';
import { JournalRoutes } from '../journal/routes/JournalRoutes';
import { CheckingAuth } from '../ui';
import { useEffect } from 'react';
import { onAuthStateChanged } from 'firebase/auth';
import { FirebaseAuth } from '../firebase/config';
import { login, logout } from '../store/auth';

export const AppRouter = () => {
  const { status } = useSelector<any, any>(state => state.auth);
  const dispatch = useDispatch<any>();
  useEffect(() => {
    onAuthStateChanged( FirebaseAuth, async (user) => {
      if(!user) return dispatch( logout(null) );
      dispatch( login(user) );
    })
  }, [])
  
  if( status === 'checking' ) {
    return <CheckingAuth />
  }

  return (
    <BrowserRouter>
      <Routes>
        {
          status === 'authenticated'
          ? <Route path='/*' element={ <JournalRoutes />} />            
          : <Route path='/auth/*' element={ <AuthRoutes />} />
        }
        <Route path='/*' element={ <Navigate to='auth/login' />} />  
      </Routes>
    </BrowserRouter>
  )
}
