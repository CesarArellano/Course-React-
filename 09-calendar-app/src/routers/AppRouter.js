import React from 'react'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect
} from "react-router-dom";

import { LoginScreen } from '../components/auth/LoginScreen';
import { CalendarScreen } from '../components/calendar/CalendarScreen';

  // exact /login => LoginScreen
  // exact /  => Calendar
export const AppRouter = () => {
  return (
    <Router>
      <Switch>
        <Route exact path="/login" component={ LoginScreen }/>
        <Route exact path="/" component={ CalendarScreen }/>
        <Redirect to="/" />
      </Switch>
    </Router>
  )
}
