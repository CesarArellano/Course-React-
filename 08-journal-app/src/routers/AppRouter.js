import React, { useEffect, useState } from 'react'
import {
  BrowserRouter as Router,
  Switch,
  Route,
} from "react-router-dom";
import { JournalScreen } from '../components/journal/JournalScreen';
import { AuthRouter } from './AuthRouter';
import { firebase } from '../firebase/firebase-config';
import { useDispatch } from 'react-redux';
import { login } from '../actions/auth';

export const AppRouter = () => {

  const dispatch = useDispatch();
  const [checking, setChecking] = useState(true);
  const [ isLoggedIn, setIsLoggedIn ] = useState(false)
  
  useEffect(() => {
    firebase.auth().onAuthStateChanged( (user) => {
      if( user?.uid ) {
        dispatch(login(user.uid, user.displayName))
        setIsLoggedIn(true);
      } else {
        setIsLoggedIn(false);
      }
      setChecking(false);
    });
  }, [dispatch, setChecking]);

  if( checking ) {
    return (
      <div className="center-complete-box">
        <div className="utils__spinner"></div>
      </div>
    )
  }
  
  return (
    <Router>
      <Switch>
        <Route path="/auth" component={ AuthRouter } />
        <Route exact path="/" component={ JournalScreen } />
      </Switch>
    </Router>
  )
}
