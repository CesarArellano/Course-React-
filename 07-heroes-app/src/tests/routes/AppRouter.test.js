import { shallow, mount } from 'enzyme';
import { AuthContext } from '../../auth/AuthContext';
import { AppRouter } from '../../routers/AppRouter';

describe('Testing in <AppRouter />', () => {
  
  test('should show login if you are not authenticated ', () => {
    const contextValue = {
      dispatch: jest.fn(),
      user: {
        logged: false,
      }
    }

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter />
      </AuthContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should show marvel if you are authenticated ', () => {
    const contextValue = {
      dispatch: jest.fn(),
      user: {
        name: 'CÃ©sar',
        logged: true,
      }
    }

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter />
      </AuthContext.Provider>
    );
    
    expect(wrapper.find('.navbar').exists()).toBe(true);
  });
})
