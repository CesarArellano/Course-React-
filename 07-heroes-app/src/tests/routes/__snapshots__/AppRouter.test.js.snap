// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <AppRouter /> should show login if you are not authenticated  1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login
    </h1>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Sign In
    </button>
  </div>
</div>
`;
